dex/fiatrates/oracle.go

```diff
@@ -0,0 +1,189 @@
+package fiatrates
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+const (
+	// FiatRateDataExpiry : Any data older than FiatRateDataExpiry will be
+	// discarded.
+	FiatRateDataExpiry = 60 * time.Minute
+
+	// averageRateRefreshInterval is how long it'll take before a fresh fiat
+	// average rate is calculated.
+	averageRateRefreshInterval = defaultRefreshInterval + 1*time.Minute
+)
+
+// Oracle manages and retrieves fiat rate information from all enabled rate
+// sources.
+type Oracle struct {
+	sources   []*source
+	ratesMtx  sync.RWMutex
+	fiatRates map[string]*FiatRateInfo
+}
+
+func NewFiatOracle(cfg Config) *Oracle {
+	return &Oracle{
+		fiatRates: make(map[string]*FiatRateInfo),
+		sources:   fiatSources(cfg),
+	}
+}
+
+// assets retrieves all assets that data can be fetched for.
+func (o *Oracle) assets() []string {
+	o.ratesMtx.Lock()
+	defer o.ratesMtx.Unlock()
+	var assets []string
+	for sym := range o.fiatRates {
+		assets = append(assets, sym)
+	}
+	return assets
+}
+
+// Rate returns the current fiat rate information for the provided symbol.
+// Returns zero if there are no rates for the provided symbol yet.
+func (o *Oracle) Rate(symbol string) float64 {
+	o.ratesMtx.Lock()
+	defer o.ratesMtx.Unlock()
+
+	symbol = parseSymbol(symbol)
+	rateInfo := o.fiatRates[symbol]
+	hasRateInfo := rateInfo != nil
+	if hasRateInfo && time.Since(rateInfo.LastUpdate) < FiatRateDataExpiry && rateInfo.Rate > 0 {
+		return rateInfo.Rate
+	}
+
+	if !hasRateInfo {
+		// Initiate an entry for this asset. Data for this asset will be fetched
+		// in the next refresh cycle.
+		o.fiatRates[symbol] = new(FiatRateInfo)
+	}
+
+	return 0
+}
+
+// Run starts goroutines that refreshes fiat rate every source.refreshInterval.
+// This should be called in a goroutine as it's blocking.
+func (o *Oracle) Run(ctx context.Context, log dex.Logger) {
+	var wg sync.WaitGroup
+	var sourcesEnabled int
+	for i := range o.sources {
+		fiatSource := o.sources[i]
+		if fiatSource.disabled {
+			log.Infof("Fiat rate source %q is disabled...", fiatSource.name)
+			continue
+		}
+
+		wg.Add(1)
+		go func(s *source) {
+			defer wg.Done()
+			ticker := time.NewTicker(s.requestInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
```

Comment:
Should this check once on start up? It waits until the first tick.

Reply:
> Does one need to request a rate for it to be added to the map?

Yes, this is to ensure the server is not fetching data that might not be used as each API call matter.

Reply:
I've enabled operators to start off with specified assets.